File: apps/angular-app/src/app/features/portfolio/domain/repositories/portfolio.repository.ts
import { Observable } from 'rxjs';
import { Result } from '../../../../core/errors/either';
import { Subscription } from '../entities/subscription.entity';

export abstract class PortfolioRepository {
  abstract getActiveSubscriptions(): Observable<Result<Subscription[]>>;
  abstract cancelSubscription(subscription: Subscription): Observable<Result<void>>;
}

File: apps/flutter-app/lib/features/portfolio/domain/repositories/portfolio_repository.dart
import 'package:dartz/dartz.dart';
import 'package:fondos_btg/core/error/failures.dart';
import 'package:fondos_btg/features/portfolio/domain/entities/subscription.dart';

abstract class PortfolioRepository {
  /// Returns all active subscriptions for the user.
  Future<Either<Failure, List<Subscription>>> getActiveSubscriptions();

  /// Cancels an active subscription.
  /// Returns the restored amount on success.
  Future<Either<Failure, double>> cancelSubscription({
    required int subscriptionId,
  });
}
